### ISSUE TITLE
404 and trailing slash with named servers

### ISSUE THREAD

**Issue by jtpio** [jupyterhub/traefik-proxy#91]
I've been trying out Traefik 1.7 as the proxy for JupyterHub (externally managed), as part of a Docker compose stack. 

It seems to be working well overall when using the default server. However redirecting to named servers give back a 404 when using the link from the JupyterHub UI:

![traefik-named-servers](https://user-images.githubusercontent.com/591645/75458710-9cd24a00-597e-11ea-97cf-be6401f1546b.gif)

The issue seems to be related to the trailing `/`. Manually adding it to the URL redirects to the server properly.

The generated rules contain the trailing `/` (one default server, and one named server called `python`):

```
[frontends.frontend__2F]
backend = "backend__2F"
passHostHeader = true

[frontends.frontend__2Fuser_2Fjtp_2F]
backend = "backend__2Fuser_2Fjtp_2F"
passHostHeader = true

[frontends.frontend__2Fuser_2Fjtp_2Fpython_2F]
backend = "backend__2Fuser_2Fjtp_2Fpython_2F"
passHostHeader = true

[frontends.frontend__2F.routes.test]
rule = "PathPrefix:/"
data = "{\"hub\": true}"

[frontends.frontend__2Fuser_2Fjtp_2F.routes.test]
rule = "PathPrefix:/user/jtp/"
data = "{\"user\": \"jtp\", \"server_name\": \"\"}"

[frontends.frontend__2Fuser_2Fjtp_2Fpython_2F.routes.test]
rule = "PathPrefix:/user/jtp/python/"
data = "{\"user\": \"jtp\", \"server_name\": \"python\"}"

[backends.backend__2F.servers.server1]
url = "http://hub:8081"
weight = 1

[backends.backend__2Fuser_2Fjtp_2F.servers.server1]
url = "http://192.168.144.4:8888"
weight = 1

[backends.backend__2Fuser_2Fjtp_2Fpython_2F.servers.server1]
url = "http://192.168.144.5:8888"
weight = 1
```

Also using the default proxy (CHP) seem to work fine.

Maybe this is something that was brought up before? Or is it most likely a configuration issue?

**Comment by jtpio** [comment]
This can also be reproduced using this more minimal config, with Traefik started by JupyterHub: https://gist.github.com/jtpio/88d98746ed8577403d4e4497370c6950

**Comment by GeorgianaElena** [comment]
Hey @jtpio! Happy you tried out traefik-proxy and I'm sorry it didn't work as it should.
I'll investigate this and come back with updates once I figure out what's wrong.
Thanks for reporting the issue!

**Comment by jtpio** [comment]
Thanks @GeorgianaElena!

**Comment by GeorgianaElena** [comment]
Hey @jtpio! I think I found the issue.

When calling `add_route`, JupyterHub ensures there's a trailing "/".
However, CHP, unlike traefik-proxy strips this slash and makes no difference between
 `/user/u1/server_name` and  `/user/u1/server_name/` when it comes to routing ([ref](https://github.com/jupyterhub/jupyterhub/blob/4e8acc71c671d39b0ae6f140d894c983bc448233/jupyterhub/proxy.py#L737)). Since this CHP behavior is marked as a bug in the comment linked in the previous reference, I don't think we should replicate it in traefik-proxy as a solution to this issue.

Instead, I believe we can solve this in JH, by adding a trailing slash to the `server_url` when there's a named server. The trailing slash exists for default servers is because `user.server_url` in this case is [`user.url`](https://github.com/jupyterhub/jupyterhub/blob/master/jupyterhub/user.py#L399) which is [`user.base_url`](https://github.com/jupyterhub/jupyterhub/blob/master/jupyterhub/user.py#L394), which [gets a trailing '/'](https://github.com/jupyterhub/jupyterhub/blob/master/jupyterhub/user.py#L163).

Sorry for making this hard to follow with all the code refs. Will open a PR in jupyterhub repo and see what people think about it.

**Comment by jtpio** [comment]
Thanks for looking into it and for the details!

Adding the trailing slash to the named servers sounds good indeed, especially if this is already the case for the default server.
